controllers/csm_controller.go

```diff
@@ -1524,12 +1529,68 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
 	return r.K8sClient
 }
 
-func (r *ContainerStorageModuleReconciler) GetMatchingNodes(ctx context.Context, labelKey string, labelValue string) (*corev1.NodeList, error) {
+func (r *ContainerStorageModuleReconciler) GetMatchingNodes(ctx context.Context, labelKey string) (*corev1.NodeList, error) {
 	nodeList := &corev1.NodeList{}
 	opts := []client.ListOption{
-		client.MatchingLabels{labelKey: labelValue},
+		client.HasLabels{labelKey},
 	}
 	err := r.List(ctx, nodeList, opts...)
 
 	return nodeList, err
 }
+
+// Invalid zoning configurations:
+// 1. Some arrays have a zone label A, but one of the arrays has zone label B, or no label
+// 2. Array secret has zone label A, but one of the nodes has zone label B, or no label
+// 3. Some of the nodes have a zone label, but arrays have no zone labels.
+// 4. All of the arrays have a zone label, but nodes have no zone label.
+func (r *ContainerStorageModuleReconciler) ZoneValidation(ctx context.Context, cr *csmv1.ContainerStorageModule, namespace string) error {
+	log := logger.GetLogger(ctx)
+	nodeList := &corev1.NodeList{}
+
+	err := r.Client.List(ctx, nodeList)
+	if err != nil {
+		return fmt.Errorf("error listing nodes %v", err)
+	}
+	numberOfNodes := len(nodeList.Items)
+
+	zones, err := drivers.ExtractZones(ctx, cr, r.Client, namespace)
+	if err != nil {
+		return fmt.Errorf("ZoneValidation failed with error: %v", err)
+	}
+
+	// Make sure all zone keys are the same, if not fail
+	var firstKeyLabel string
+	for key := range zones {
+		if firstKeyLabel == "" {
+			firstKeyLabel = key
+			continue
+		}
+		if key != firstKeyLabel {
+			return fmt.Errorf("detected mismatched Keys in zones")
```

Comment:
As a user, I would like to know more specifically where is the misconfiguration: e.g. which array gave the error, or maybe a hint similar to "check the array info secret, zone key should be the same for all arrays".